napari/config/core.py

```diff
@@ -0,0 +1,708 @@
+"""napari configuration
+
+This module is copied, with modifications, from dask.config
+
+We like Dask's "plain python dict" treatment of configurations, handling of
+nested namespaces, backwards compatilibity of new settings, context management
+(temporary setting) of config values, choice of yaml for persistence, and
+environment variable options
+
+see documentation for dask.config at:
+https://docs.dask.org/en/latest/configuration.html
+
+Configuration is specified in one of the following ways:
+
+1. YAML files in ~/.config/napari/ or /etc/napari/
+2. Environment variables that start with ``NAPARI_``
+3. Default settings within sub-libraries
+
+This combination makes it easy to specify configuration in a variety of
+settings ranging from personal workstations, to IT-mandated configuration, to
+docker images.
+
+Loading
+-------
+
+When this module is imported, this sequence of events happens:
+
+1. the config is cleared and ``refresh()`` is called
+2. ``refresh()`` updates the empty config with values from dicts in the
+   `config.defaults` list (currently empty)
+3. ``refresh()`` then updates the config by calling ``collect()`` ...
+   a. ``collect()`` gathers settings from various places:
+      - ``collect_yaml()`` looks for *any* `.yaml` files in the
+         ``config.paths`` list.  Notably, this includes
+         ``~/.config/napari/napari.yaml`` and anything else in there.
+      - ``collec_env()`` then looks for any environment variables that begin
+         with ``NAPARI_``
+   b. finally, ``collect()`` calls `merge()` on all of those discovered dicts,
+      to yield one merged result and returns it to the ``refresh()`` function
+      (by default, dicts discovered later in the process take priority).
+4. back in the ``refresh()`` function, that final merged result is used to
+   update the global config dict.
+5. At that point, the ``napari.yaml`` in the repo is read, if one doesn't
+   already exist, a file is created at ``~/.config/napari/napari.yaml`` (with
+   ``ensure_file()``) that is an exact copy of the repo ``napari.yaml`` but
+   with everything commented out, as a way to show users what they can change.
```

Comment:
This seems to come really late in the game? Shouldn't this be the first thing that happens?

Reply:
it doesn't really matter where `ensure_file` is called.  it's just creating a commented-out copy of the `napari.yaml` that ships with the repo at `~/.config/napari/napari.yaml` if one doesn't already exist there (so curious users can see what options are available to them).  It doesn't have any effect on the settings that are loaded.